wgsl/extract-grammar.py

```diff
@@ -255,118 +338,287 @@ def parse(lines, i):
+            elif production["type"] == "token":
+                if production["value"] in result[scanner_token.name()]:
+                    subsource = f"<a for=syntax_sym lt={result[scanner_token.name()][production['value']]}>`{production['value']}`</a>"
+                else:
+                    subsource = f"`{production['value']}`"
+            elif production["type"] == "pattern":
+                subsource = f"`{production['value']}`"
+            return subsource
+        elif production["type"] == "sequence":
+            if len(rule_name) == 0:
+                return (
+                    "( "
+                    + " ".join(
+                        [
+                            reproduce_rule_source(member)
+                            for member in production["value"]
+                        ]
+                    )
+                    + " )"
+                )
+            else:
+                return " ".join(
+                    [reproduce_rule_source(member) for member in production["value"]]
+                )
+        elif production["type"] == "choice":
+            return (
+                "( "
+                + " | ".join(
+                    [reproduce_rule_source(member) for member in production["value"]]
+                )
+                + " )"
+            )
+        elif production["type"] == "match_0_1":
+            if len(production["value"]) == 1:
+                return reproduce_rule_source(production["value"][0]) + " ?"
+            else:
+                return (
+                    "( "
+                    + " ".join(
+                        [
+                            reproduce_rule_source(member)
+                            for member in production["value"]
+                        ]
+                    )
+                    + " ) ?"
+                )
+        elif production["type"] == "match_1_n":
+            if len(production["value"]) == 1:
+                return reproduce_rule_source(production["value"][0]) + " +"
+            else:
+                return (
+                    "( "
+                    + " ".join(
+                        [
+                            reproduce_rule_source(member)
+                            for member in production["value"]
+                        ]
+                    )
+                    + " ) +"
+                )
+        elif production["type"] == "match_0_n":
+            if len(production["value"]) == 1:
+                return reproduce_rule_source(production["value"][0]) + " *"
+            else:
+                return (
+                    "( "
+                    + " ".join(
+                        [
+                            reproduce_rule_source(member)
+                            for member in production["value"]
+                        ]
+                    )
+                    + " ) *"
+                )
+        else:
+            print(production["type"])
```

Comment:
Should this raise an exception?  It happens when there's a missing case in this method.
Also, it does not have to be in the "else" clause.

Reply:
That is correct